File: built-in plugins dsp/mb-comp-mastering.cpp
/**
 * MB Mastering Comp
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Transparent mastering compressor
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_COMP_MASTERING_H
#define MB_COMP_MASTERING_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCompMastering : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-comp-mastering";
    static constexpr const char* PLUGIN_NAME    = "MB Mastering Comp";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -8f;  // range [-30, 0]
    float ratio = 1.5f;  // range [1, 4]
    float attack = 20f;  // range [1, 100]
    float release = 250f;  // range [50, 1000]
    };

    MbCompMastering() = default;
    ~MbCompMastering() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -30f, 0f);
        params.ratio = std::clamp(params.ratio, 1f, 4f);
        params.attack = std::clamp(params.attack, 1f, 100f);
        params.release = std::clamp(params.release, 50f, 1000f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Mastering Comp
        return input;
    }
};

#endif // MB_COMP_MASTERING_H

File: built-in plugins dsp/mb-comp-mastering.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCompMasteringPlugin: PluginDefinition = { id: 'mb-comp-mastering', slug: 'mb-comp-mastering', name: 'MB Mastering Comp', category: 'effect', type: 'compressor', version: '1.0.0', description: 'Transparent mastering compressor', author: 'Max Booster', grade: 'A', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -8, minValue: -30, maxValue: 0, automatable: true }, { id: 'ratio', name: 'Ratio', type: 'float', defaultValue: 1.5, minValue: 1, maxValue: 4, automatable: true }, { id: 'attack', name: 'Attack', type: 'float', defaultValue: 20, minValue: 1, maxValue: 100, automatable: true }, { id: 'release', name: 'Release', type: 'float', defaultValue: 250, minValue: 50, maxValue: 1000, automatable: true }], defaultPreset: { threshold: -8, ratio: 1.5, attack: 20, release: 250 } };

export default MbCompMasteringPlugin;
